[GTK][WPE] Add ConnectionGLib

carlosgcampos opened the pull request:
#### 0d195d406a9f3e43668024ffff6b0ca79f7ac56c
<pre>
[GTK][WPE] Add ConnectionGLib
<a href="https://bugs.webkit.org/show_bug.cgi?id=297897">https://bugs.webkit.org/show_bug.cgi?id=297897</a>

Reviewed by Michael Catanzaro.

Add a Connection implementation using GLib sockets API and simplifying
the code by stop supporting stream sockets.

Canonical link: <a href="https://commits.webkit.org/299386@main">https://commits.webkit.org/299386@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/470c5c3aba36076c06a09fc0daeaac1c78481f74

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124800 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70680 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38998 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46884 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90029 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59570 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121574 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31066 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106352 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70534 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30125 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24463 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68458 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110739 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24654 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127859 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117135 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34355 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98667 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102573 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98451 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43894 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21890 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42032 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18930 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45398 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51076 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145831 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44861 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37494 "Found 1 new JSC binary failure: testapi, Found 18593 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48208 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46548 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,408 @@
+    {
+        zeroBytes(*this);
+        *this = info;
+    }
```
```suggestion
    AttachmentInfo()
    {
        // The entire AttachmentInfo is passed to write(), so we have to zero our
        // padding bytes to avoid writing uninitialized memory.
        zeroBytes(*this);
    }

    AttachmentInfo(const AttachmentInfo& info)
        : AttachmentInfo()
    {
        *this = info;
    }
```

spenap commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,408 @@
+    auto& messageInfo = outputMessage.messageInfo();
+    const auto& attachments = outputMessage.attachments();
+    GOutputVector outputVector[3];
+    int outputVectorLenth = 0;
```
typo: 
```suggestion
    int outputVectorLength = 0;
```

mcatanzaro replied:
Carlos either likes Ctrl+V or autocomplete. It's hopeless to expect that typo won't be precisely the same in every single location. :)

carlosgcampos replied:
I love middle click paste :-D

spenap commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,408 @@
+                attachmentInfo[i].setNull();
+        }
+        if (!fds.isEmpty()) {
+            // Use g_unix_fd_message_new_with_fd_list() to create the message wihtout duplicating the file descriptors.
```
```suggestion
            // Use g_unix_fd_message_new_with_fd_list() to create the message without duplicating the file descriptors.
```

mcatanzaro commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -43,26 +43,17 @@
-#include <wtf/glib/GUniquePtr.h>
-#endif
-
 #if OS(DARWIN)
 #define MSG_NOSIGNAL 0
 #endif
```
```suggestion
```

Probably safe to delete this because: I think only WebKitGTK built this file on Darwin, and now that will never happen anymore. And if I'm wrong, then the build should fail, somebody should complain, and we can just put it back.

mcatanzaro replied:
(Feel free to ignore this if you don't want to change it.)

carlosgcampos replied:
Yes, I thought that if someone is building WebKitGTK on darwing they could just pick ConnectionUnix at build time.

mcatanzaro commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -43,26 +43,17 @@
 #include <wtf/TZoneMallocInlines.h>
 #include <wtf/UniStdExtras.h>
 
-#if USE(GLIB)
-#include <gio/gio.h>
-#include <wtf/glib/GUniquePtr.h>
-#endif
-
 #if OS(DARWIN)
 #define MSG_NOSIGNAL 0
 #endif
 
 // Although it's available on Darwin, SOCK_SEQPACKET seems to work differently
-// than in traditional Unix so fallback to STREAM on that platform.
+// than in traditional Unix so fallback to DGRAM on that platform.
 #if defined(SOCK_SEQPACKET) && !OS(DARWIN)
 #define SOCKET_TYPE SOCK_SEQPACKET
 #else
-#if USE(GLIB)
-#define SOCKET_TYPE SOCK_STREAM
-#else
 #define SOCKET_TYPE SOCK_DGRAM
 #endif
-#endif // SOCK_SEQPACKET
```
I think you can delete all this too, for the same reason.

`SOCKET_TYPE` is notably only used in one place, and there it should always be `SOCK_DGRAM` because the GLib and Darwin conditions probably won't ever be true after this. All other calls to `createPlatformConnection` use hardcoded `SOCK_SEQPACKET`, so you could also remove the socket type parameter of `createPlatformConnection` entirely if you want to.

mcatanzaro replied:
(Feel free to ignore this if you don't want to change it.)

mcatanzaro commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,408 @@
+#include <wtf/glib/GRefPtr.h>
+#include <wtf/glib/GUniquePtr.h>
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
```
I don't expect you to address -Wunsafe-buffer-usage warnings in this pull request, but perhaps you could narrow them down to particular functions or lines of code, rather than suppressing them throughout the entire file?

carlosgcampos replied:
I just copied this from ConnectionUnix. How can I know if there are unsafe buffer accesses? Does building with default options fail? or only in the EWS safer-cpp bot fails?

mcatanzaro replied:
Building with Clang would fail. Our EWS might be using GCC.

mcatanzaro reviewed (approved):
Hmmm, I'm not enthusiastic about the level of code duplication here, but after considering various ways to share some code in certain places, e.g. `AttachmentInfo` or `createMessageDecoder`/`processMessage` or `sendOutgoingMessage`, I'm pretty sure it wouldn't actually simplify anything. Most of the functions are slightly different in each file, and trying to share between the two files will probably make things more complicated rather than cleaner. At least they are both fairly short. So I'd say what you have here is fine.

aperezdc commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,576 @@
+#if OS(ANDROID)
+    Type m_type;
+#endif
+
```
This empty line could be removed.

aperezdc commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,576 @@
+        auto& attachmentInfo = consumeAndReinterpretCastTo<AttachmentInfo>(messageData);
+        size_t attachmentIndex = attachmentCount - i - 1;
+#if OS(ANDROID)
+        switch (attachmentInfo[i].type()) {
```
It looks like you carried over the indexing from `ConnectionUnix.cpp`:

```suggestion
        switch (attachmentInfo.type()) {
```

aperezdc commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,576 @@
+                attachments[attachmentIndex] = WTFMove(m_fileDescriptors[fdIndex++]);
+            break;
+        case AttachmentInfo::Type::HardwareBuffer:
+            if (attachmentInfo[i].isNull())
```
Same here:

```suggestion
            if (attachmentInfo.isNull())
```

aperezdc commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,576 @@
+        if (!receiveIncomingHardwareBuffers())
+            return;
+
+        if (auto decoder = createMessageDecoder(m_readBuffer, m_fileDescriptors))
```
This method does not take any parameters:

```suggestion
        if (auto decoder = createMessageDecoder())
```

aperezdc commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,576 @@
+        for (size_t i = 0; i < attachments.size(); ++i) {
+#if OS(ANDROID)
+            RELEASE_ASSERT(attachmentInfo[i].type() == AttachmentInfo::Type::Unset);
+            switchOn(attachment[i],
```
Typo:

```suggestion
            switchOn(attachments[i],
```

aperezdc reviewed (changes requested):
There's a few typos in the Android bits, please see the comments.

aperezdc commented on Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp:
```diff
@@ -0,0 +1,575 @@
+                [&](const UnixFileDescriptor& fd) {
+                    attachmentInfo[i].setType(AttachmentInfo::Type::FileDescriptor);
+                    if (fd)
+                        fds.append(attachments[i].value());
```
This should use `fd`:

```suggestion
                        fds.append(fd.value());
```

aperezdc reviewed (changes requested):
One more fix needed for Android...

carlosgcampos committed:
[GTK][WPE] Add ConnectionGLib
https://bugs.webkit.org/show_bug.cgi?id=297897

Reviewed by Michael Catanzaro.

Add a Connection implementation using GLib sockets API and simplifying
the code by stop supporting stream sockets.

Canonical link: https://commits.webkit.org/299386@main
